### PR TITLE
Add uuid to default user collection location

### PR DESCRIPTION
**Pull request by ellio167**
This PR changes the way the kim-api handles (by default) the user collection configuration files and directories.

It now uses a UUID to identify each separate installation on the system so that separate installations do not need to share user collections.

The issue and suggestion for implementation comes from Axel:
https://github.com/lammps/lammps/pull/1440#issuecomment-515398685
https://github.com/lammps/lammps/pull/1440#issuecomment-515536747

### CODE REVIEW

**Comment by ellio167** [comment]
@tadmor @yafshar @dskarls I'd like to be sure that we are all OK with this change and see if there are any suggestions for improvement.

**Comment by yafshar** [comment]
> @tadmor @yafshar @dskarls I'd like to be sure that we are all OK with this change and see if there are any suggestions for improvement.



> @tadmor @yafshar @dskarls I'd like to be sure that we are all OK with this change and see if there are any suggestions for improvement.

@ellio167 It is an excellent feature to have. It enables users to have multiple versions with different compilers.
Till now my solution for having multiple versions was through a shell script based on compiler name and KIM version
```sh
mkdir -p ~/.kim-api/${kim}_${cmp}/model-drivers-dir
mkdir -p ~/.kim-api/${kim}_${cmp}/portable-models-dir
mkdir -p ~/.kim-api/${kim}_${cmp}/simulator-models-dir
```

and then using another script to set the config file for each version as,
```sh
kim_config="~/.kim-api/${kim}_${cmp}"
base="install_dir"

$base/bin/kim-api-collections-management set-user-model-drivers-dir    $kim_config/model-drivers-dir"
$base/bin/kim-api-collections-management set-user-portable-models-dir  $kim_config/portable-models-dir"
$base/bin/kim-api-collections-management set-user-simulator-models-dir $kim_config/simulator-models-dir"
```

To sum it up, it is a very nice feature to have. 

- Issue
  If you have multiple versions on the system, with the current approach of creating `UUID`, it is tough to find out which one is outdated to remove it.
- Suggestion
  Maybe change the `UUID` to have some hint of the system, compilers, or kim_version to make it readable by the user as well. You can make it similar to `KIM_ID` or something like `${KIMVERSION}_${CXX}_${CC}_${FC}_${UUID}`

**Comment by dskarls** [comment]
> * Issue
>   If you have multiple versions on the system, with the current approach of creating `UUID`, it is tough to find out which one is outdated to remove it.
> * Suggestion
>   Maybe change the `UUID` to have some hint of the system, compilers, or kim_version to make it readable by the user as well. You can make it similar to `KIM_ID` or something like `${KIMVERSION}_${CXX}_${CC}_${FC}_${UUID}`
> 
> I see that you provide the installation information in a README file within a created folder, but again I think it is nicer to make the name unique & human readable.

Are you suggesting there's a way to have cmake append such information to the README file (which currently just lists the KIM API install directory)?  I saw Axel also proposed having some indication of compiler, etc, in his comments but surely the source code itself should never be aware of that.  Actually, this direction also seems to be redundant with pkg-config.  Is pkg-config capable (or, better yet, good) at handling concurrent installations of a package?

File: CMakeLists.txt [hunk 1]
```diff
@@ -109,8 +116,8 @@ include(DefaultCompilerFlags)
 
 if(MINGW)
   # On MinGW Windows platform, make <dlfcn.h> functions (dlopen, dlsym, etc.) available
-  # using a third-party portability layer: https://github.com/dlfcn-win32/dlfcn-win32 
-  # It should be installed in the MSYS2 environemnt using the command: 
+  # using a third-party portability layer: https://github.com/dlfcn-win32/dlfcn-win32
+  # It should be installed in the MSYS2 environemnt using the command:
```
Comment (dskarls) [inline CMakeLists.txt:120]: typo: "environemnt" -> "environment"

**Comment by yafshar** [comment]
> > * Issue
> >   If you have multiple versions on the system, with the current approach of creating `UUID`, it is tough to find out which one is outdated to remove it.
> > * Suggestion
> >   Maybe change the `UUID` to have some hint of the system, compilers, or kim_version to make it readable by the user as well. You can make it similar to `KIM_ID` or something like `${KIMVERSION}_${CXX}_${CC}_${FC}_${UUID}`
> > 
> > I see that you provide the installation information in a README file within a created folder, but again I think it is nicer to make the name unique & human readable.
> 
> Are you suggesting there's a way to have cmake append such information to the README file (which currently just lists the KIM API install directory)? I saw Axel also proposed having some indication of compiler, etc, in his comments but surely the source code itself should never be aware of that. Actually, this direction also seems to be redundant with pkg-config. Is pkg-config capable (or, better yet, good) at handling concurrent installations of a package?

@dskarls No not in the README file. It would be nice to change the FOLDER (user collection location) which is currently set to UUID to a readable name similar to the KIM items, with the human-readable part which tells the user which Kim version or which compilers are used to build the specific version and also some random part (or date or anything else) to make it unique.

**Comment by yafshar** [comment]
@dskarls  For example, I have made these in my setup,
```bash                    
2.1.3_clang_gcc9   
2.1.3_clang_gcc10                 
2.1.3_gcc9                                
2.1.3_gcc10                              
2.1.3_intel_2020_U1            
2.2.0_clang_gcc9                
2.2.0_clang_gcc10                  
2.2.0_gcc9                            
2.2.0_gcc10                 
2.2.0_intel_2020_U1 
```
It makes it easy to choose which model I want to use or delete from the system.

**Comment by ellio167** [comment]
These are good suggestions.  Since the values that are used here are set at configure time, it's not really a problem to obtain and embed some compiler info (at least in the dir names).

I can see how a human-readable id is more desirable than the uuid.  I was using the timestamp for configuration as the seed to the uuid anyway; I can just use it directly to ensure uniqueness.

File: CMakeLists.txt (hunk 1 above)
Comment (ellio167) [reply CMakeLists.txt:120]: Thanks; resolved elsewhere...

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,6 +63,10 @@ mark_as_advanced(KIM_API_ENABLE_SANITIZE)
 option(KIM_API_ENABLE_COVERAGE "Enable code coverage" OFF)
 mark_as_advanced(KIM_API_ENABLE_COVERAGE)
 #
+string(TIMESTAMP _KIM_API_CONFIGURATION_TIMESTAMP "%Y-%m-%d-%H-%M-%S" UTC)
+set(KIM_API_CONFIGURATION_TIMESTAMP "${_KIM_API_CONFIGURATION_TIMESTAMP}" CACHE STRING "UTC imestamp for configuration of kim-api")
```
Comment (dskarls) [inline CMakeLists.txt:67]: typo: 'imestamp' -> 'timestamp'

**Comment by ellio167** [comment]
OK.  I've used the full semver version string with prerelease and build metadata.  (Previously the build metadata only included the `${CMAKE_CXX_COMPILER_ID}` value, now I've added also the `C` and `Fortran` values.)  To the end of the version string I've appended a UTC timestamp which should be sufficiently unique.

This produces a pretty long string when building from the git repo with a dirty workingdir (e.g.,
`2.2.0-git+v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`) but it's not unmanageable and it should be much shorter for an actual release build.

FYI, I'll parse out the string:
`2.2.0-git+v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`

`2.2.0` -- is the semver Major, Minor, and Patch version
`-` -- is the separator for the prerelease
`git` -- is the prerelease component of the version
`+` -- is the separator for the build metadata
`v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53` -- is the build metadata component of the version

Parsing the build metadata:
`v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`

`v2.1.3-git-92-gce05f22a.dirty` -- is the output of `git describe --dirty=.dirty --broken=.broken --always` (where `v2.1.3-git` is a tag in this case, there are 92 commits since that tag, and `gce05f22a` is the hash of the last commit)
`AppleClang` -- is the CXX compiler ID
`AppleClang` -- is the C compiler ID
`GNU` -- is the Fortran compiler ID
`2020-07-22-02-09-53` -- is the timestamp `YYYY-MM-DD-HH-MM-SS` in UTC

File: CMakeLists.txt (hunk 2 above)
Comment (ellio167) [reply CMakeLists.txt:67]: Thanks

**Comment by yafshar** [comment]
> OK. I've used the full semver version string with prerelease and build metadata. (Previously the build metadata only included the `${CMAKE_CXX_COMPILER_ID}` value, now I've added also the `C` and `Fortran` values.) To the end of the version string I've appended a UTC timestamp which should be sufficiently unique.
> 
> This produces a pretty long string when building from the git repo with a dirty workingdir (e.g.,
> `2.2.0-git+v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`) but it's not unmanageable and it should be much shorter for an actual release build.
> 
> FYI, I'll parse out the string:
> `2.2.0-git+v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`
> 
> `2.2.0` -- is the semver Major, Minor, and Patch version
> `-` -- is the separator for the prerelease
> `git` -- is the prerelease component of the version
> `+` -- is the separator for the build metadata
> `v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53` -- is the build metadata component of the version
> 
> Parsing the build metadata:
> `v2.1.3-git-92-gce05f22a.dirty.AppleClang.AppleClang.GNU.2020-07-22-02-09-53`
> 
> `v2.1.3-git-92-gce05f22a.dirty` -- is the output of `git describe --dirty=.dirty --broken=.broken --always` (where `v2.1.3-git` is a tag in this case, there are 92 commits since that tag, and `gce05f22a` is the hash of the last commit)
> `AppleClang` -- is the CXX compiler ID
> `AppleClang` -- is the C compiler ID
> `GNU` -- is the Fortran compiler ID
> `2020-07-22-02-09-53` -- is the timestamp `YYYY-MM-DD-HH-MM-SS` in UTC

@ellio167 The name is a bit long, but it is explicit and has all the information one might need.
```sh
2.2.0-git+v2.1.3-git-93-g0c5cf509.GNU.GNU.GNU.2020-07-22-12-18-54
2.2.0-git+v2.1.3-git-93-g0c5cf509.Intel.Intel.Intel.2020-07-22-12-22-25
2.2.0-git+v2.1.3-git-93-g0c5cf509.AppleClang.AppleClang.GNU.2020-07-22-12-30-04
```
- Suggestion
  One extra suggestion is to add a user-defined folder name maybe by something like `-DUserCollectionName=${user_defined_name}`, which can override the default one.

**Comment by ellio167** [comment]
@yafshar Yes, there are settings to change all of these defaults at configure time:

`KIM_API_CONFIGURATION_TIMESTAMP`
`KIM_API_USER_CONFIGURATION_FILE` -- note this specifies the name of the *file* not just the directory
`KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT`
`KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT`
`KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT`

I've edited the `INSTALL` file with updated descriptions of these.

**Comment by ellio167** [comment]
> @yafshar Yes, there are settings to change all of these defaults at configure time:
> 
> `KIM_API_CONFIGURATION_TIMESTAMP`
> `KIM_API_USER_CONFIGURATION_FILE` -- note this specifies the name of the _file_ not just the directory
> `KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT`
> `KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT`
> `KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT`
> 
> I've edited the `INSTALL` file with updated descriptions of these.

Simplified these a bit:
`KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT` -> `KIM_API_USER_MODEL_DRIVERS_DIR_DEFAULT`
`KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT` ->`KIM_API_USER_PORTABLE_MODELS_DIR_DEFAULT`
`KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT` -> `KIM_API_USER_SIMULATOR_MODELS_DIR_DEFAULT`

**Comment by yafshar** [comment]
> > @yafshar Yes, there are settings to change all of these defaults at configure time:
> > `KIM_API_CONFIGURATION_TIMESTAMP`
> > `KIM_API_USER_CONFIGURATION_FILE` -- note this specifies the name of the _file_ not just the directory
> > `KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT`
> > `KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT`
> > `KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT`
> > I've edited the `INSTALL` file with updated descriptions of these.
> 
> Simplified these a bit:
> `KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT` -> `KIM_API_USER_MODEL_DRIVERS_DIR_DEFAULT`
> `KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT` ->`KIM_API_USER_PORTABLE_MODELS_DIR_DEFAULT`
> `KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT` -> `KIM_API_USER_SIMULATOR_MODELS_DIR_DEFAULT`

@ellio167 sounds great. I think there is a typo in INSTALL file, [KIM_API_CONFIGURATION_FILE](https://github.com/openkim/kim-api/blob/d68cbe53614806ef8e47e0c4eea4bab834f6fd36/INSTALL#L187) should be `KIM_API_USER_CONFIGURATION_FILE`.

**Comment by ellio167** [comment]
> @ellio167 sounds great. I think there is a typo in INSTALL file, [KIM_API_CONFIGURATION_FILE](https://github.com/openkim/kim-api/blob/d68cbe53614806ef8e47e0c4eea4bab834f6fd36/INSTALL#L187) should be `KIM_API_USER_CONFIGURATION_FILE`.

Actually, it is correct (but maybe confusing), the environment variable is named `KIM_API_CONFIGURATION_FILE` but the cmake configuration setting is `KIM_API_USER_CONFIGURATION_FILE`

**Comment by yafshar** [comment]
> `KIM_API_CONFIGURATION_FILE`

@ellio167 OK it was not clear for me from the text. (maybe we can add an example like the rest of the installation doc where you provided examples"

**Comment by yafshar** [comment]
> @yafshar Yes, there are settings to change all of these defaults at configure time:
> 
> `KIM_API_CONFIGURATION_TIMESTAMP`
> `KIM_API_USER_CONFIGURATION_FILE` -- note this specifies the name of the _file_ not just the directory
> `KIM_API_USER_MODEL_DRIVER_PLURAL_DIR_DEFAULT`
> `KIM_API_USER_PORTABLE_MODEL_PLURAL_DIR_DEFAULT`
> `KIM_API_USER_SIMULATOR_MODEL_PLURAL_DIR_DEFAULT`
> 
> I've edited the `INSTALL` file with updated descriptions of these.

@ellio167 This approach seems very strange to me. Even when one has the environment variable `KIM_API_CONFIGURATION_FILE`, then based on the current approach it creates two folders. One based on the `UID` which is the default one and the second one based on `KIM_API_CONFIGURATION_FILE` where it only contains a config file and a README file. In the second case, isn't it better to replace the `UID` with the new location when we have an environment variable or similar to the traditional prefix approach in configuration (where you can define the installation location), one can also decide on a folder name?

**Comment by ellio167** [comment]
@yaser Typically a user would create their own configuration file if they are going to use the `KIM_API_CONFIGURATION_FILE` environment variable.  There is no reason to expect that the directories listed in this configuration file would be in subdirectories of the directory in which the configuration file exists.  However, if for some reason the file does not exist, the kim-api creates it with defaults and creates those default directories.

Your comment does make me realize that it is not a good idea to create the README.txt because of the above.  I'll plan to refactor things so that the content of the README.txt file comes at the end of the configuration file.

**Comment by yafshar** [comment]
@ellio167 thanks, the explanation in the `INSTALL` file makes it clear.

**Pull request merged**